soc/raspberrypi/CMakeLists.txt

```diff
@@ -1,3 +1,3 @@
 # SPDX-License-Identifier: Apache-2.0
 
-add_subdirectory(${SOC_SERIES})
+add_subdirectory(rp2xxx)
```

Comment:
I've just started looking at it, so my opinion may change later, but it looks like it would be better to separate it into rp2040 and rp2350. It seems like it would be better to have the same configuration as pico-sdk.

Reply:
I disagree. Splitting the directory here into rp2040 and rp2050 seems to result in a lot of identical or almost identical files in sibling directories. I feel like it's easier to see the differences when they're closer together, e.g. soc/raspberrypi/rp2xxx/Kconfig.defconfig makes it clear that for all rp2xxx devices `NUM_IRQS` is defined as a function of the SoC series.

Reply:
Here, priority should be given to mapping the group configuration defined by the vendor as is rather than file deduplication.
In other words, rp2350 is needed as the series that encompasses the 235[04][AB] SoC, and rpi_pico (or rp2 would be fine) is necessary as the family that includes the rp2350 series and rp2040 series.
In other words, Zephyr's free is very limited because the vendor's series configuration is like this.

Reply:
I reviewed the current configuration; it would be better to split the directory here, as I mentioned.

Since we separated 2040 and 2350 by SOC_SERIES, we can fit it to the typical configuration used by Zephyr,
`soc/${manufacturer}/${soc_family}/${soc_series}/Kconfig.defconfig.${soc}`
format.

In other words, I think it would be better to fit it like this:
soc/raspberrypi/rpi_pico/rp2350/Kconfig.defconfig.rp2350a

If there are a lot of duplicate files, create
soc/raspberrypi/rpi_pico/common
is a good idea.

Reply:
It may be persistent, but I request again to split 20xx into 2040 and 2350 here.

One reason is that, as mentioned earlier, many HWMv2s have this type of structure.

The other is that these two are essentially different SoCs, and the BSP code is completely separated.
In other words, the code on the surface is similar, but structurally, they are different, so they should have different structures.

In fact, soc.c also has

#if CONFIG_PLATFORM_SPECIFIC_INIT

void z_arm_platform_init(void)
{
	runtime_init_per_core_enable_coprocessors();
}

#endif /* CONFIG_PLATFORM_SPECIFIC_INIT */`

```

, but this is used instead of dividing it into 2040/2350, so we think the file structure itself should be separated.

Sorry, but I would appreciate your opinion. > @nordicjm @yonsch

Reply:
The two SoCs are completely different, they should be separate files

Reply:
> It may be persistent, but I request again to split 20xx into 2040 and 2350 here.

@soburi, please can I ask that you allow other people the opportunity to contribute to the conversation. Replying to your own replies to keep bumping the same request is not great.

That said, to answer you comments/concerns:

> ...these two are essentially different SoCs, and the BSP code is completely separated.

We know that's not true. There's plenty of common code in the Pico SDK where the source code is common (the same *.c file), and SoC(/SoC-family) specific elements are chosen at pre-processor time. If they really were "completely separated", there would be two almost identical sibling directories in the SDK.

> In fact, soc.c also has
> ``` 
>     #if CONFIG_PLATFORM_SPECIFIC_INIT
> ```

Regardless of where this is in the Zephyr directory structure, it's something that a user may want to turn of at build-configuration time. The co-processors aren't required at all when using the Hazard3 cores, for example, or if the user wants to turn of the double-precision FPU (perhaps for benchmarking, power saving, it's not Zephyr's remit to know/control). This is also consistent with other parts of the same file, I like consistency :-).
```
#if CONFIG_RP2_REQUIRES_SECOND_STAGE_BOOT
```

> The two SoCs are completely different

Many of the peripherals are drop-in compatible as-is, or have minor differences.

Who has the final say on this in the review process? The codeowner is @yonsch , AIUI.

Reply:
Note: my comment is from the little knowledge I have of this which it is 2x cortex M33's and/or 2x RISCV cores, other than that I have no knowledge on this part, internal or external. No other SoC in zephyr combines multiple generations with different architecture in the SoC files, they might use common drivers or common HALs but the SoC files are per device

Reply:
@nordicjm 

At the moment, I don't have a clear idea about RISC-V support, but I think it's realistic to use "revision.cmake".

When RISC-V and ARM cores exist simultaneously, it's a completely unknown area.
It's an exciting topic, but it's obvious that the discussion will have no convergence, so it's better to focus on pico2 as a "better pico" here, I think.

Reply:
@ajf58 

> > ...these two are essentially different SoCs, and the BSP code is completely separated.
> 
> We know that's not true. There's plenty of common code in the Pico SDK where the source code is common (the same *.c file), and SoC(/SoC-family) specific elements are chosen at pre-processor time. If they really were "completely separated", there would be two almost identical sibling directories in the SDK.

That's right. So we need to respect the NOT commonized part also as well as the commonized one.
The soc parts have been split into rp2040 and rp2350, also in the pico-sdk.
Creating a directory for each SoC model is the majority structure for Zephyr, and it is also the same as the pico-sdk structure.
There is no need to go to the trouble of integrating them into one.

> > In fact, soc.c also has
> > ```
> >     #if CONFIG_PLATFORM_SPECIFIC_INIT
> > ```
> 
> Regardless of where this is in the Zephyr directory structure, it's something that a user may want to turn of at build-configuration time. The co-processors aren't required at all when using the Hazard3 cores, for example, or if the user wants to turn of the double-precision FPU (perhaps for benchmarking, power saving, it's not Zephyr's remit to know/control). This is also consistent with other parts of the same file, I like consistency :-).


If this is the basis, there should be this option to enable/disable it for RP2040 as well, but in reality, this is not the case.
`runtime_init_per_core_enable_coprocessors` is an M33-dependent implementation and does not exist in rp2040 builds, creating an apparent asymmetry here.
In other words, this consistency is closed in RP2350.
I think this fact rather encourages splitting RP2040 and RP2350 for reasons of consistency.


> Many of the peripherals are drop-in compatible as-is, or have minor differences.

Yes, it is. These are well-supported and commonized in the driver layer.

> Who has the final say on this in the review process? The codeowner is @yonsch , AIUI.

There is no "arbitrator" here; only consensus is needed.
If I had to say, it would be the TSC, but that also requires consensus building.

@yonsch is the maintainer, and I act as a co-maintainer, so if @yonsch and I agree on something, whether positive or negative, things go smoothly, but if not, we start to have a long discussion.
So far, there have been no major divergences in our policies.

Reply:
> @nordicjm
> 
> At the moment, I don't have a clear idea about RISC-V support, but I think it's realistic to use "revision.cmake".
> 
> When RISC-V and ARM cores exist simultaneously, it's a completely unknown area. It's an exciting topic, but it's obvious that the discussion will have no convergence, so it's better to focus on pico2 as a "better pico" here, I think.

No, it would need to use CPU clusters defined at SoC level, board revisions are for board revisisions

Reply:
@nordicjm

> No, it would need to use CPU clusters defined at SoC level, board revisions are for board revisisions

Ah, Ok.
It seems an essential problem, so this obstacle will not be solved quickly.
I create #78222 to discuss this problem.

Reply:
This seems to be blocking the PR progressing, so I'm going to change it, making use of a `common` directory.